[Bookings] Add "Sort by" UI to booking list screen

irfano opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the Linear issue this PR addresses, e.g., WOOMOB-373. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR introduces “Sort by” for the Bookings list. When tapped, it opens a bottom sheet. This PR only covers the UI implementation.

The selected “Sort by” option will be persisted across tabs.

i2 design: lxixW6aBBNkCd9ooBK70Z1-fi-359_5652

### Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->
1. Use a CIAB site.
2. Manually install the bookings plugin from this ZIP file: p1758531103469849/1758101141.913349-slack-C03L1NF1EA3
3. You'll need to enable bookings v2 on your CIAB testing site. You can do that by installing the Code [Snippets plugin](https://wordpress.org/plugins/code-snippets/) to your site and adding the following PHP snippet: `define( 'WC_BOOKINGS_NEXT_ENABLED', true );`
4. Navigate to Bookings.
5. Tap "Sort by" button

### The tests that have been performed
<!-- To give the reviewer an idea of what could be missed in terms of testing -->
Screens below

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->
<img width="360" alt="Screenshot_20251002_190719" src="https://github.com/user-attachments/assets/2c555d22-c873-46e3-944b-3b227dfad1ad" /> <img width="360" alt="Screenshot_20251002_190712" src="https://github.com/user-attachments/assets/e1313150-c483-4409-8483-8c134bd0d908" />


- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

irfano committed:
Add controls state to BookingListViewState

This change introduces `BookingListControlsState` to `BookingListViewState`
to handle sort and filter actions.

irfano committed:
Add BookingListControlsState function to the BookingListViewModel

irfano committed:
Add "Sort by" and "Filter" buttons to the bookings screen

irfano committed:
Add string resources for "Sort by" bottom sheet

irfano committed:
Add checkmark icon

irfano committed:
Add booking sort bottom sheet

This commit introduces a new bottom sheet for sorting bookings.
The bottom sheet allows users to sort bookings by newest to oldest or oldest to newest.

irfano committed:
Handle bookings list sorting in view model

irfano committed:
Add booking list sort options into the view state model

irfano committed:
Show sort bottom sheet when visible

irfano committed:
Add booking list sorting tests

Added tests to verify the sort functionality in the booking list.

irfano commented:
> ...Do you want to change the UI before I start on the review?

Got it! I missed that. I'l update it. Please hold off on reviewing until my change is ready.

irfano committed:
Merge branch 'trunk' into issue/WOOMOB-1237-booking-list-sorting-ui

irfano committed:
Wire selected sort option into controls state and ViewModel

irfano committed:
Add sort label helpers and strings

irfano committed:
Use OutlinedButtons in controls and show sort label with arrow

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/bookings/list/BookingListScreen.kt:
```diff
@@ -71,6 +82,10 @@ fun BookingListScreen(state: BookingListViewState) {
+            if (state.contentState.isNotEmpty()) {
+                BookingListControls(state.controlsState)
+                HorizontalDivider()
+            }
```
I wonder if we should move this to `BookingList` inside the pull to refresh box 🤔 

Currently the pull to refresh indicator behavior is different from the other tabs:

| Products | Bookings |
| --- | --- |
| <img width="1080" height="2400" alt="Image" src="https://github.com/user-attachments/assets/23bc46c9-77d4-4c9b-8196-95357e6fd80a" /> | <img width="1080" height="2400" alt="Image" src="https://github.com/user-attachments/assets/81725bf0-6fbb-410f-9b69-3d4d7289b5df" /> |

hichamboushaba replied:
With the below change 592ce86, I think this won't be straightforward to change without changing the whole screen organization, so I think we can keep it as it's now.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/bookings/list/BookingListScreen.kt:
```diff
@@ -152,6 +167,51 @@ private fun BookingList(
+            .padding(horizontal = 8.dp, vertical = 4.dp),
+        horizontalArrangement = Arrangement.spacedBy(8.dp)
+    ) {
+        OutlinedButton(
```
WDYT about using `AssistedChip` for this instead of a button? It seems to me chips are more suited for this.

If you agree, maybe we can even create a simple component `WCChip` that sets the correct theming, mainly the `shape` that should be `CircleShape` and the the min size that you set.

irfano replied:
It’s good advice. I tried it, but unfortunately, I couldn’t set a minimum width and center the label at the same time. I want to set a minimum width to match the buttons on the Products screen. If you look at the Filters button, you’ll notice it has a minimum width, so the current approach is more flexible. We also use outlined buttons in `products_sort_and_filters_card.xml`. 

So, let's keep it as it is.

<img width="533" height="291" alt="image" src="https://github.com/user-attachments/assets/e108ebd6-c325-42d9-82a8-5664b8f20043" />

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/bookings/list/BookingListScreen.kt:
```diff
@@ -71,6 +82,10 @@ fun BookingListScreen(state: BookingListViewState) {
             )
+            if (state.contentState.isNotEmpty()) {
+                BookingListControls(state.controlsState)
+                HorizontalDivider()
```
To my eye, the divider we use in the products screen is less subtle than what we have now, and I think it's because for the other screen it's not a divider, it's a shadow.

Do you think it's better to set this divider's thickness to `0.5dp` to look similar?

irfano replied:
Goot catch! Done: c748a1fd36093eaf9c945bbaaf9bdeaaebbad538

hichamboushaba reviewed (approved):
Nice work @irfano 💯, I left some suggestions, please check them out, but I'm pre-approving as it works as expected.

irfano committed:
Always show booking list controls

The `BookingListControls` and `HorizontalDivider` are now always displayed on the booking list screen, regardless of whether there is content to show.

irfano commented:
I also made a change to set these new control buttons as always visible  (592ce86b1cf288a473ec25c3dc893cd2c257d81b). We can’t rely on `state.contentState.isNotEmpty()` for button visibility, since the content could be empty due to the filter state. Therefore, the filter should always remain visible.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/bookings/list/BookingListViewModel.kt:
```diff
@@ -26,6 +26,16 @@ class BookingListViewModel @Inject constructor(
     private val loadingState = MutableStateFlow(BookingListLoadingState.Idle)
     private val selectedTab = savedStateHandle.getStateFlow(viewModelScope, BookingListTab.Today)
 
+    private val sortOptionsByTab = MutableStateFlow(
+        mapOf(
+            BookingListTab.Today to BookingListSortOption.NewestToOldest,
+            BookingListTab.Upcoming to BookingListSortOption.NewestToOldest,
+            BookingListTab.All to BookingListSortOption.NewestToOldest,
+        )
+    )
```
@irfano I missed this during my first review, do we need to have different sorting behavior for each tab? personally I don't think it's needed, and actually as a user it could be confusing to me why the sort I selected is lost when I switch tab (I won't know it's kept for the other tab), so I suggest we simplify this and keep only a single value for the whole screen, WDYT?

irfano replied:
I have two reasons why I think it should be kept per tab:
1. In the design, the “Sort by” and “Filter” buttons are placed below the tab bar, which makes them feel like they belong to the tab itself rather than the parent **Bookings** screen.
2. In the sync call, we discussed that date filters won’t be available for some tabs. This means filter options won’t be the same across all tabs. If we used a single filter selection across tabs, a user might set a date filter and not realize that it isn’t available in other tabs.

hichamboushaba replied:
> In the sync call, we discussed that date filters won’t be available for some tabs. This means filter options won’t be the same across all tabs. If we used a single filter selection across tabs, a user might set a date filter and not realize that it isn’t available in other tabs.

Not necessarly, we can probably keep the other filters and exclude just the date filter when switching from `All` to the other tabs, at least that's what I thought we'll do from the discussion.

Anyway, personally the experience of changing the sorting when switching tabs was confusing, and as a user I didn't expect it to behave this way, so I think a product decision is needed, and also to make sure iOS follows the same thing. Can you ask Wagner and Katie about this? I'm OK with merging the PR as it's, and address if needed in a separate PR.

irfano replied:
Asked here: p1759505490361729-slack-C09FHQNQERG

> I'm OK with merging the PR as it's, and address if needed in a separate PR.

Sounds good! I'm merging this.

irfano replied:
We decided to persist the “Sort by” option across tabs. I’ve made that change: 4c7d1dea1f096c8e2e752f324228c8f2af8f552f

irfano committed:
Merge branch 'trunk' into issue/WOOMOB-1237-booking-list-sorting-ui

# Conflicts:
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/bookings/list/BookingListScreen.kt
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/bookings/list/BookingListViewModel.kt
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/bookings/list/BookingListViewState.kt
#	WooCommerce/src/main/res/values/strings.xml
#	WooCommerce/src/test/kotlin/com/woocommerce/android/ui/bookings/list/BookingListViewModelTest.kt

irfano committed:
Fix divider thickness in BookingListScreen

The thickness of the HorizontalDivider in the BookingListScreen has been explicitly set to 0.5.dp to ensure a thinner appearance.

irfano committed:
Merge branch 'trunk' into issue/WOOMOB-1237-booking-list-sorting-ui
